worlds/soe/__init__.py

```diff
@@ -170,31 +174,23 @@ class SoEWorld(World):
     location_name_to_id, location_id_to_raw = _get_location_mapping()
     item_name_groups = _get_item_grouping()
 
-    trap_types = [name[12:] for name in option_definitions if name.startswith('trap_chance_')]
+    trap_types = [name[12:] for name in options_dataclass.__dataclass_fields__ if name.startswith('trap_chance_')]
```

Comment:
not sure about this one. is this fine or do we have a better solution? @alwaysintreble

Reply:
idk that there's really a better way, but it'd be a bit cleaner to put this directly on the dataclass instead of doing it here?

Reply:
I think I might try something like this (not tested):

`options.py`
```python3
@dataclass
class SoEOptions(PerGameCommonOptions):
    difficulty:            Difficulty
    energy_core:           EnergyCore
    ...

    def trap_chances(self) -> Iterator[TrapChance]:
        for field in fields(self):
            if issubclass(field.type, TrapChance):
                yield getattr(self, field.name)
```
---

`__init__.py`
```python3
        trap_names: typing.List[str] = []
        trap_weights: typing.List[int] = []
        if trap_count > 0:
            for trap_option in self.options.trap_chances():
                trap_names.append(trap_option.item_name)
                trap_weights.append(trap_option.value)
            if sum(trap_weights) == 0:
                trap_weights = [1 for _ in trap_weights]

        def create_trap() -> Item:
            return self.create_item(self.random.choices(trap_names, trap_weights)[0])
```

Reply:
Interestingly this fails with
`TypeError: issubclass() arg 1 must be a class`
for
> accessibility's .type is Accessibility, which is <class 'str'>

Pretty sure that's because of `from __future__ import annotations` in /Options.py, but field.type becomes a sad thing if it doesn't handle future annotations implicitly. :/


```python
if isinstance(field.type, type) and issubclass(field.type, ...):
```
seems kinda jank, but I don't have a better idea. Do you?

Reply:
Yeah, that's sad about `field.type` - I wonder if Python code devs are aware of that...

In that case, I might do this:

```python3
    def trap_chances(self) -> Iterator[TrapChance]:
        for field in fields(self):
            option = getattr(self, field.name)
            if isinstance(option, TrapChance):
                yield option
```